### PR TITLE
feat(#692): add configurable budget limit with percentage-used tracking

### LINKED ISSUE #692: Feature request: 7-day daily cost chart with budget tracking
This feature is needed by users running multi-agent Hermes swarms. A reference implementation exists in the [Hermes Control Interface repo](https://github.com/xaspx/hermes-control-interface). Please add a bar chart showing daily API spend over the last 7 days, alongside a monthly pace projection and a configurable budget limit with percentage used. This is essential for users self-hosting Hermes with cost-sensitive API providers.

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- The cost chart already computes a monthly pace projection from 7-day snapshots. The natural extension is to let the user anchor that projection against a self-defined budget, no new data sources, no new backend infrastructure.
- The budget is one more setting stored in `settings.json`, surfaced in the `/api/provider/cost-history` response as `monthly_budget` so the frontend can compute the percentage inline without a second round-trip. The existing `limit` field stays as the OpenRouter provider credit limit.
- Pace-vs-budget is the right display metric, not month-to-date actual, because the 7-day snapshot window can't reliably reconstruct full-month spend, but the pace projection already answers "am I on track to stay under budget?"

## What Changed

- **`api/config.py`**: added `provider_cost_budget` to `_SETTINGS_DEFAULTS`, then centralized validation in `_coerce_provider_cost_budget(...)` so save-time persistence rejects zero, negative, non-finite, over-cap, and rounded-to-zero sub-cent values before they can be stored.
- **`api/providers.py`**: `_get_provider_cost_budget()` now reuses that same helper on read, so manually edited out-of-range values no longer leak back out through `monthly_budget`; `get_provider_cost_history` still includes `monthly_budget` in all three OpenRouter response paths (`no_key`, `unavailable`, `available`) while leaving `limit` as the provider credit limit.
- **`static/panels.js`**: `renderProviderCostChart` annotates the Monthly Pace label with `(N% of budget)` when a budget is configured, amber at `>=80%` and red at `>=100%`; new `_attachBudgetControls` appends a budget input row with Set/Clear buttons and a progress bar with color-coded fill.
- **`static/style.css`**: added compact CSS for the budget row, input group, progress bar, and pace annotation coloring.
- **`static/i18n.js`**: added six new keys in `en`, with the existing locale fallback chain covering the rest.
- **`tests/test_provider_cost_budget.py`**: covers settings persistence, endpoint response shape, frontend wiring, the `0.004` rounded-to-zero regression, and the read-side cap for manually edited settings.

## Why It Matters

Users running OpenRouter for production workloads can now compare their projected monthly spend against a self-defined budget directly on the quota card. "Monthly pace: $142.80 (142% of budget)" turns a number into a decision. The feature stays opt-in, users without a budget configured see no change.

## Verification

- `pytest tests/test_provider_cost_budget.py -v --timeout=60`
- `pytest tests/ -v --timeout=60`
- Manual: open Settings -> Providers. Confirm the "Monthly budget" input appears below the cost chart. Set a budget, verify the pace annotation shows `(N% of budget)` and a progress bar appears. Clear the budget, verify the progress bar disappears. Refresh usage, confirm the chart and budget control re-render correctly.

## Upstream

Closes #692.

## Model Used

Claude Opus 4.6 via Claude Code CLI

### CODE REVIEW

File: static/panels.js [hunk 1]
```diff
@@ -9361,7 +9369,103 @@ async function renderProviderCostChart(card){
...
+    }catch(e){
+      if(typeof showToast==='function') showToast(t('provider_cost_budget_save_failed'));
+    }
+  }
+
+  setBtn.addEventListener('click',()=>{
+    const val=parseFloat(input.value);
+    if(!isFinite(val)||val<=0) return;
+    _saveBudget(val);
+  });
```
[the remark this answers is not in the document]
Comment (rodboev) [reply static/panels.js:9464]: The backend validation is the source of truth and correctly rejects >= 1e9. The frontend validates the basics (finite, positive) while the server owns the full constraint. A user entering a billion-dollar budget sees it silently revert to the prior value, which is safe behavior. Adding the redundant upper-bound mirror to the frontend would mean maintaining the same constant in two places.

File: api/config.py [hunk 2]
```diff
@@ -8057,6 +8058,18 @@ def save_settings(settings: dict) -> dict:
...
                 v = cleaned
+            if k == "provider_cost_budget":
+                if v is None or v == "":
+                    current[k] = None
+                    continue
+                try:
+                    v = float(v)
+                except (TypeError, ValueError):
+                    continue
+                if not (0 < v < 1e9) or not __import__("math").isfinite(v):
```
[the remark this answers is not in the document]
Comment (rodboev) [reply api/config.py]: Technically correct that Python's chained comparison short-circuits before reaching isfinite() for NaN/Inf. The guard is kept as a defensive belt-and-suspenders measure that makes the rejection of non-finite values explicit in the code rather than relying on an implicit property of chained comparisons.

File: static/panels.js [hunk 3]
```diff
@@ -9361,7 +9369,103 @@ async function renderProviderCostChart(card){
...
+  const titleDiv=document.createElement('div');
+  titleDiv.className='provider-cost-budget-title';
+  titleDiv.textContent=t('provider_cost_budget_label');
+  row.appendChild(titleDiv);
+
+  const inputGroup=document.createElement('div');
+  inputGroup.className='provider-cost-budget-input-group';
+
+  const prefix=document.createElement('span');
+  prefix.className='provider-cost-budget-prefix';
```
[the remark this answers is not in the document]
Comment (rodboev) [reply static/panels.js:9401]: The remove-then-rerender pattern follows the existing codebase convention for card refreshes (e.g. _refreshProviderQuota replaces the entire card via card.replaceWith). The settings POST is awaited before removal, so the wrap is only removed after a successful save. If the subsequent history fetch inside renderProviderCostChart fails, the chart recovers on the next panel reload.

[the remark this answers is not in the document]
Comment (rodboev) [reply static/panels.js:9401]: > Shall I remember this as a rule: "accept remove-then-rerender as the codebase convention for card refreshes; don't flag it as a gap unless there's actual data loss or an unrecoverable user-visible error without a recovery path"?

@nesquena-hermes ?

**Comment by nesquena-hermes** [comment]
## 🔬 Gate certification — RED ⛔ (two narrow validation fixes away)

**Certified head:** `sha:59f3e27329e3` · **PR:** #5120 · _rodboev — "feat(#692): add configurable budget limit with percentage-used tracking"_
**Verdict:** Strong, well-tested feature — robust validation, all-locale i18n parity, XSS-safe DOM, 23 passing tests, live API contract verified. But the gate (on a clean rebase) found **two real, reproduced edge cases** in the budget validation itself, one of which makes a user's budget silently vanish. Both fixes are trivial; fix-then-merge.

### What I ran (isolated worktree `/tmp/wt-gate-5120`, gate port 8783)
| Gate | Result |
|------|--------|
| Codex (reproduce, on clean rebase) | **SHIP ONLY WITH FIXES** — 2 verified edge cases (below) |
| Opus (full review) | **COMMENT** — no blockers; 1 minor fire-and-forget re-render note + 3 cosmetic |
| Full pytest suite (rebased head `aee358cf`) | **11105 passed, 0 failed** (green — but does not cover the two edge cases below) |
| PR's own tests (clean checkout) | **23 passed** |
| In-process validation probe | 13/13 real cases correct (rejects 0/neg/NaN/inf/≥1e9/invalid-string) |
| Live API contract | `GET /api/provider/cost-history?provider=openrouter` returns `monthly_budget` field ✓ |
| i18n parity | budget keys present in all **13 locales** ✓ |
| Browser | Insights renders, `_attachBudgetControls`/`renderProviderCostChart` wired, **0 JS errors** |

### ⚠️ Stale-base note (resolved)
Codex's first pass returned KICK BACK with 13 CORE/SILENT findings — **all stale-base artifacts** from other shipped PRs (Codex itself said "No budget-specific blocker found; the blockers are current-master regressions introduced by this branch delta"). I rebased onto current master (`537ff28a`) locally — clean (`rc=0`), post-rebase diff is exactly the 6 budget files, mergeable. Re-running Codex on the clean diff surfaced the two real findings below.

### ⛔ Blocking (both reproduced by me)
**1. `api/config.py` save_settings — sub-cent budget silently lost.** A tiny positive value passes the `0 < v < 1e9` check, then `round(v, 2)` collapses it to `0.0`, which is stored. Cost-history treats `0.0` as unset (`if val > 0`), so the budget **silently disappears**. Reproduced:
```
budget=0.004 -> stored=0.0   (silently lost)
budget=0.001 -> stored=0.0   (silently lost)
budget=0.009 -> stored=0.01  (ok)
```
**Fix:** validate the *rounded* value before storing — `rounded = round(v, 2); if rounded <= 0: continue; current[k] = rounded`.

**2. `api/providers.py` `_get_provider_cost_budget()` — missing upper-bound on read.** It only checks `val > 0 and isfinite`, not the `< 1e9` cap `save_settings` enforces. A manually-edited `settings.json` (or any value that bypassed save) is emitted unbounded. Reproduced:
```
manual settings budget=5e12 -> _get_provider_cost_budget()=5000000000000.0  (no upper-bound guard)
```
**Fix:** apply the same guard on read — `0 < val < 1e9 and math.isfinite(val)` — ideally via a shared validator used by both save and read (removes the duplication that let them drift).

### Findings (the feature is otherwise solid)
- ✅ **Validation is rigorous** for the common cases (0/negative/NaN/inf/≥1e9/invalid-string all rejected — 13/13 in-process). The two gaps above are the rounding-to-zero boundary and the read-side bound.
- ✅ **i18n parity** across all 13 locales (head commit added budget keys everywhere).
- ✅ **API contract verified live** — `monthly_budget` threaded into all 3 cost-history response shapes (no_key/unavailable/success).
- ✅ **XSS-safe DOM** (Opus-confirmed), percentage calc sound (`pct = round(pace/budget*100)` with warn≥80% / over≥100% classes).
- ⚠️ Opus non-blocking: the save handler's re-render is fire-and-forget (data saves; only the control display is lost until refresh until the await resolves). Worth a small await/guard but degrades gracefully.

### Recommendation to the next agent
**Bounce for the two validation fixes (both trivial, both reproduced), then it's a clean merge.** This is a quality feature from a T1 contributor — not a rejection. Ideal fix folds both into a single shared budget-validator used by `save_settings` and `_get_provider_cost_budget` so they can't drift. After that, re-gate (full suite + a regression test for `0.004 → rejected/clamped` and the read-side cap). I applied `gate-fail`; I did not apply `changes-requested` (warm-up owns auto-bounce). Cert valid only at `sha:59f3e27329e3`.

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy/close. Cert valid only at `sha:59f3e27329e3`; a new push invalidates it → re-gate._

**Comment by rodboev** [comment]
Applied the two gate fixes in `04d0cde8`.

1. `api/config.py` now routes `provider_cost_budget` through a shared `_coerce_provider_cost_budget(...)` helper before persisting it, so values like `0.004` are rejected instead of being rounded down to stored `0.0` and disappearing from the chart.
2. `api/providers.py` now reuses that same helper on read, so manually edited out-of-range values no longer surface through `monthly_budget`.

I also added regressions in `tests/test_provider_cost_budget.py` for the rounded-to-zero case and the read-side cap.

**Comment by nesquena-hermes** [comment]
## 🔬 Gate certification — GREEN ✅ (re-cert: prior gate-fail fixed)

**Certified head:** `sha:04d0cde8430c` · **PR:** #5120 · _rodboev — "feat(#692): add configurable budget limit with percentage-used tracking"_
**Verdict:** The two issues that blocked my earlier gate-fail (`59f3e273`) are **both fixed** in this re-push, exactly as recommended — via a shared `_coerce_provider_cost_budget` validator used by save AND read. Full gate now green: Codex SAFE, Opus APPROVE, full suite + 25 PR tests pass, both fixes verified in-process. Ready to merge.

### What I ran (isolated worktree `/tmp/wt-gate-5120b`)
| Gate | Result |
|------|--------|
| Codex (reproduce, on clean rebase) | **SAFE TO SHIP** (no findings; 40/40 targeted budget tests + node-check) |
| Opus (full review) | **APPROVE** — both prior blockers fixed; 3 cosmetic non-blocking nits |
| Full pytest suite (`-p no:xdist`, rebased head `0196bb6c`) | **11109 passed, 0 failed** (to completion) |
| PR's own tests (clean checkout) | **25 passed** (+2 new regression tests for the fixed gaps) |
| In-process gap-fix verification (mine) | **both gaps fixed** — see below |

### The fix (addresses my prior gate-fail `comment-4829342731`)
The re-push (`04d0cde8`, commit "keep budget validation consistent across save and read") adds a **shared validator** and routes both paths through it — exactly the "single shared validator so they can't drift" fix I recommended:
```python
def _coerce_provider_cost_budget(value):
    try: rounded = round(float(value), 2)
    except (TypeError, ValueError): return None
    if not (0 < rounded < 1e9) or not math.isfinite(rounded): return None
    return rounded
```
- `save_settings` → uses it (rounds **before** validating).
- `_get_provider_cost_budget` (read) → now `return _coerce_provider_cost_budget(raw)` (was a divergent inline check).

### Both prior gaps verified fixed (in-process, against this head)
- ✅ **Gap 1 (sub-cent silent loss):** `save_settings({"provider_cost_budget": 0.004})` → stored `None` (rejected), not `0.0`. The round-before-validate ordering means a value that would collapse to `0.0` now fails `0 < rounded` and is rejected — no silent disappearance.
- ✅ **Gap 2 (read-side missing cap):** a manually-edited `settings.json` with `5e12` → `_get_provider_cost_budget()` returns `None` (the `< 1e9` cap is now enforced on read via the shared validator).
- ✅ Valid value (`50.5`) still persists correctly.
- ✅ The PR added the **2 regression tests I asked for by name**: `test_save_settings_rejects_subcent_budget_after_rounding`, `test_get_provider_cost_budget_rejects_huge_manual_setting`.

### Findings (clean)
- ✅ Shared validator removes the duplication that caused the save/read drift — the root-cause fix, not a patch.
- ✅ i18n parity preserved (budget keys in all locales; English-fallback convention per the repo).
- ✅ Live API contract (`monthly_budget` in all 3 cost-history response shapes) intact from the prior gate.
- ⚠️ Non-blocking (Opus): `panels.js` save handler is fire-and-forget re-render (POST completes; only the visual re-render isn't awaited — degrades gracefully). Optional `await`/disable-while-inflight.
- ⚠️ Cosmetic (Opus): the `or not math.isfinite(rounded)` is technically redundant (`0 < rounded < 1e9` already excludes NaN/inf) but harmless.

### Stale-base note
Re-push was based on an older master (diff vs current master ballooned to 95 files / 7704 deletions = stale-base tell). I rebased onto current master (`3b120e70`) locally — clean (`rc=0`, 5 commits), post-rebase diff is exactly the 6 budget files, mergeable. Re-ran Codex + the full suite on the rebased head (`0196bb6c`) — both green. Rebase before merge.

### Recommendation to the next agent
**Ready to merge — cert fresh for `sha:04d0cde8430c`.** This is a fixed-and-returned PR: I gate-failed it for 2 validation gaps, rodboev fixed both with the exact shared-validator approach + the 2 regression tests, and it's now clean (Codex SAFE, Opus APPROVE, 11109/0 suite + 25/25 PR tests, both gaps empirically re-verified closed). Rebases cleanly onto current master (6-file diff). The 3 remaining notes are cosmetic/optional. The stale `gate-fail` label is from the old head — this GREEN cert supersedes it. Cert valid only at `sha:04d0cde8430c`.

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy/close. Cert valid only at `sha:04d0cde8430c`; a new push invalidates it → re-gate._

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.777** (via release PR #5293). Thanks @rodboev! 🎉 Screenshot-approved (budget field + Set/Clear + %-used progress bar); Codex SAFE + Opus APPROVE, full suite green, validated on both save+read paths. (The one red CI shard was a known TTL-cache flake in test_issue2941_wiki_browse, unrelated — cleared on re-run.)
